Feature/code reviews

Rosscrowley opened the pull request:
I have added headers for the Code Review topic.

Rosscrowley committed:
Code-Reviews-added-headers-to-be-expanded-on

Paddyjh reviewed (changes requested):
I think we should hold off on merging until we have content we are happy with for each of the topics.

Isaac811 commented:
Fair play. I agree


Best regards

Jiayu Liu


Patrick Haughey ***@***.***>于2024年3月20日 周三16:45写道：

> ***@***.**** requested changes on this pull request.
>
> I think we should hold off on merging until we have content we are happy
> with for each of the topics.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Paddyjh/Software-Quality-Assurance-Group-10/pull/3#pullrequestreview-1949483826>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARJDY6Y5KC355ZAWIARZPHLYZG4KRAVCNFSM6AAAAABE72IIXSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSNBZGQ4DGOBSGY>
> .
> You are receiving this because you were assigned.Message ID:
> <Paddyjh/Software-Quality-Assurance-Group-10/pull/3/review/1949483826@
> github.com>
>

Rosscrowley committed:
new-image-folder

Rosscrowley committed:
Creating-a-new-branch-to-create-Image-Folder

Rosscrowley committed:
Code-Reviews-first-two-topics-first-draft

Rosscrowley committed:
New-Image_Folder

Rosscrowley committed:
Code-Review-added-third-topic-and-new-images

Paddyjh committed:
Code-Reviews-fixed-headers-formatting

Changed to match develop branch

Paddyjh committed:
Merge branch 'develop' into feature/Code-Reviews

Paddyjh committed:
Code-Reviews-advantages-disadvantages-of-code-reviews

Added Advantages and Disadvantages of code reviews

Added Mitigations for each disadvantage

Paddyjh committed:
Code-Reviews-code-reviews-in-CI-CD

Added another subtopic on code reviews in CI/CD.

This improves the flow of the document as it lends itself nicely to the next topic.

This improves the flow of the document as it lends itself nicely to the next topic

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -1,9 +1,82 @@
-# Quality Assurance Handbook
+# **Quality Assurance Handbook**
```
I think we should remove the bold syntax; since headers are already bold, the "**" isn't needed. Removing this should make it easier when merging into main

Rosscrowley replied:
I agree @Paddyjh,  will make that change, thanks.

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -1,9 +1,82 @@
+ - A code review is an essential practice to ensure that the code is high quality, aligns with predetermined standards, contains no mistakes
+ - Many companies practice different processes of how to carry out code reviews. Different processes depend on a number of factors such as, the level of experience of developers, specific project or regulatory requirements,and the complexity and lifecycle stage of the project, etc. Below are some different approaches:
+    - Checklist-Based Reviews - where reviewers follow a predetermined list of checks to follow while reviewing code.
+    - Pair Programming - where a developer codes and gets real-time feedback (code review) on their code.
```
I think we could clarify that this involves two coders programming together, which is where the real-time feedback comes from. Although maybe this is implied enough by the title?

Rosscrowley replied:
I agree, I will alter the sentence slightly to make it clearer. Thanks.

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -1,9 +1,82 @@
+    - Reviewee must be receptive to feedback. They must have a willingness to learn and understand criticism is not personal.
+    - Code must be reviewed by the reviewee before requesting a code review, to save the reviewers from having to respond to miscellaneous errors such as spelling mistakes.
+    - Seek and give clarity when needed. Be prepared to offer clarity on code you have produced and seek clarity when feedback is unclear.
+    - Small, incremental code changes should be the aim as code with many lines is more straneous to review. 
```
Spelling mistake: "straneous" should be "strenuous"

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -1,9 +1,82 @@
+- As well as key practices, included in this section are some practices that we feel are beneficial to fostering a positive and effective feedback 
+
+- For Reviewers
+    - The reviewer should be experienced enough, so that they know what they are looking for such as design, logic, and bugs
```
To make this sentence flow better, we could rephrase it to something like "... what they are looking for, such as issues relating to design, logic and bugs".

Rosscrowley replied:
Agreed, will reword the sentence.

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -1,9 +1,82 @@
+        - Avoid leaving many comments where possible as this can demotivate the reviewer, instead prioritise essential comments, which brings the code up to a level of acceptance.
+        - As well as leaving constructive feedback, always attempt to leave a positive comment on something you felt was done well. This can boost the confidence of the reviewee.
+    - Reviewee
+        - Seems simple but ensure you reply to comments left by a reviewer as well as thanking them for taking the time to review your code.
```
Maybe changing the start of this to "It seems simple ..." could make this flow better. Alternatively we could just Start the sentence at "Ensure you reply ..."

Paddyjh reviewed (changes requested):
Left a few suggestions and corrections. Any questions just lemme know !

Rosscrowley committed:
Code-Review-recommened-changes-made

Isaac811 commented on Handbook/QA Handbook.md:
```diff
@@ -6,4 +6,77 @@
+    - Code reviews should not be rushed to avoid code smells entering the main codebase. Limiting code review sessions or setting a maximum line of code (LOC) per hour can ensure that the reviewer does not lose focus
+    - Feedback should be constructive. Providing constructive feedback helps to enhance understanding and learning rather then incite competition or conflict.
+- For Reviewees
+    - Reviewee must be receptive to feedback. They must have a willingness to learn and understand criticism is not personal.
```
Maybe we could change this to “code must be reviewed and tested by the reviewee” to ensure code is also fully tested before a code review.

Rosscrowley replied:
I will add that in. thanks Isaac!

Rosscrowley replied:
updated in latest commit

Isaac811 replied:
The Reviewers might spell wrong?

Rosscrowley replied:
Apologies, I see my mistake, I will correct now! Thanks.

Rosscrowley replied:
Corrected spelling mistake in latest commit!

Isaac811 commented on Handbook/QA Handbook.md:
```diff
@@ -6,4 +6,77 @@
+    - Checklist-Based Reviews - where reviewers follow a predetermined list of checks to follow while reviewing code.
+    - Pair Programming - involves two developers, where a developer codes and gets real-time feedback off another developer (code review) on their code.
+    - Pre-Commit Reviews - This is where code is reviewed before it is merged into the main branch.
+- Google mandates code reviews and and follows the Pre-commit review. Below is a diagram of their Code Review Flow.
```
Remove the double “and” in this sentence.

Rosscrowley replied:
Thanks for spotting this Isaac!

Rosscrowley replied:
updated in latest commit

Isaac811 reviewed (changes requested):
Let me know if you have a questoins , left 2 comments there. Thanks!

Rosscrowley committed:
Code-Review-further-recommended-changes-made

Rosscrowley committed:
Code-Review-corrected-spelling-mistake

Paddyjh reviewed (changes requested):
Everything looks good but i think we can introduce a standard height limit of 500px for diagrams to better organize the overall structure and readability of the document. 

This can be done by using a html image tag and specifying the height rather than the way we are currently doing it. I have attached an example below :

img src="name_of_image.png" height="500"

What do yous think ?

Rosscrowley commented:
> Everything looks good but i think we can introduce a standard height limit of 500px for diagrams to better organize the overall structure and readability of the document.
> 
> This can be done by using a html image tag and specifying the height rather than the way we are currently doing it. I have attached an example below :
> 
> img src="name_of_image.png" height="500"
> 
> What do yous think ?

@Paddyjh I agree, I will reformat the images. Thanks!

Rosscrowley committed:
Code-Review-reformatted-images

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -6,4 +6,77 @@
 
+### What Are Code Reviews?
+ 
+ - A code review is the task of someone who is not the author reviewing a piece of code before it is merged to the main codebase 
```
I think we could rephrase this a bit more clearly

Rosscrowley replied:
Sounds good, how about, "A code review is the process of getting feedback and approval from other developers on submitted code before it is merged into the main codebase."

Paddyjh replied:
That's perfect, thanks!

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -6,4 +6,77 @@
+### What Are Code Reviews?
+ 
+ - A code review is the task of someone who is not the author reviewing a piece of code before it is merged to the main codebase 
+ - A code review is an essential practice to ensure that the code is high quality, aligns with predetermined standards, contains no mistakes
```
small thing, but maybe we could change this to " and contains no mistakes."

Paddyjh commented on Handbook/QA Handbook.md:
```diff
@@ -6,4 +6,77 @@
+
+### Best Practices for Conducting Effective Code Reviews
+
+- Although there are many different approaches to conducting a code review, there are some key practices that should be considered for the code review to be effective. 
```
Could merge these two sentences into one ?

Paddyjh reviewed (changes requested):
Just some small changes before we can merge, thanks

Rosscrowley committed:
Code-Review-changes-resolving-comments

Paddyjh reviewed (approved):
Looks great, thanks for all the work on this guys!
